proposals/js-string-builtins/Overview.md

```diff
@@ -171,11 +171,11 @@ The simplest way is to have a module import each string as an immutable global.
 
 This proposal adds an extension to JS-API compile routine to support optimized 'imported string constants' to address this use-case.
 
-The `WebAssemblyCompileOptions` is extended with a `boolean importedStringConstants` flag. When this is set, the module may define imports of the form `(import "'" "%stringConstant%"" (global externref))`, and the JS-API will use the provided `%stringConstant%` import field name to be the value of the global. This allows for any UTF-8 string to be imported with minimal overhead.
+The `WebAssemblyCompileOptions` is extended with a `boolean importedStringConstants` flag. When this is set, the module may define imports of the form `(import "'" "%stringConstant%"" (global (ref extern)))`, and the JS-API will use the provided `%stringConstant%` import field name to be the value of the global. This allows for any UTF-8 string to be imported with minimal overhead.
 
 The string namespace is chosen to be the single quote ASCII character `'`. We may revise this to be a longer name before this proposal is finalized.
 
-All imports that reference this namespace must be globals of type immutable externref. If they are not, an eager compile error is emitted.
+All imports that reference this namespace must be immutable globals and have a type that matches non-nullable externref (i.e. `(ref extern)` and `externref` are both allowed). If they are not, an eager compile error is emitted.
```

Comment:
```suggestion
All imports that reference this namespace must be immutable globals and have type non-nullable externref (i.e. `(ref extern)`). If they are not, an eager compile error is emitted.
```
`(ref null extern)` isn't a subtype of `(ref extern)`, it's the other way round. And there's probably no reason to allow both? Hosts that support js-string-builtins populate these globals with non-null values, and for hosts that don't, supplying `""` as a default import value is no harder than supplying `null`.

Reply:
I was imagining that when this option is enabled, the JS-API creates a `(global (ref extern))` for each string constant that is the provided value for whatever the user asks to import (as if it was provided by the imports object).

The global import checking rules already allow for an immutable global import to be a supertype of the actual provided value. e.g. a module importing `(global anyref)` can be instantiated with a `(global eqref)` value. This only works for immutable globals.

I agree there's probably not much use for it, but I don't want to have special rules here so I thought I'd just match how it works already.

Reply:
I'm not arguing for special rules, and I'm aware that a "slot" (global, local, param, etc) of type X can be satisfied with any value whose type is a subtype of X, and I'm not saying that nullable `externref` wouldn't work.

Line 174 says
> the module may define imports of the form `(import "'" "%stringConstant%"" (global (ref extern)))`

with the type just being tightened to non-nullability in this PR. Line 178 then saying 
> `(ref extern)` and `externref` are both allowed

seems to be at odds with that. I'm suggesting that we pick the simple option: the type of the imports must be `(ref extern)`, the type of the values satisfying them must be `(ref extern)`, subtyping works as it always does (which is trivial in this case). We certainly could certainly allow more freedom, but if we get _zero_ benefit from it, why not be conservative and keep things as simple as possible?

Reply:
One benefit of allowing magic imports to work anywhere normal imports would is that it would allow producers that don't know about GC and non-nullable references (e.g. LLVM) emit code that uses string builtins and magic imports. In fact, LLVM could emit such code today with no modifications as long as the globals are allowed to have nullable types.

Reply:
What do you mean by "allowing magic imports to work anywhere normal imports would"?

(I also don't really see how a legacy module producer that hasn't been taught about this proposal would accidentally generate string constants in the form of imported globals that just so happen to fit this proposal's requirements, with or without nullability.)

Reply:
@jakobkummerow 

Agree that line 174 and 178 seem contradictory. I reworked it to try to make it more clear.

I might be missing something here, but I think this current approach is the simplest. We're basically specifying that the JS-API creates globals of a specific type: `immutable (ref extern)`, and provides it to the [normal import checking logic](https://webassembly.github.io/gc/core/valid/matching.html#match-globaltype). The user can specify any imported global type as long as it matches with what the JS-API creates. In this case that allows for either non-nullable and nullable, but requires immutable.

If we wanted to only allow for `ref extern`, we'd need to create a new matching rule that checks for immutable and and also an exact valtype match. Not a big deal, but this approach just seemed to follow the spec better.

Reply:
> We're basically specifying that the JS-API creates globals of a specific type: `immutable (ref extern)`, and provides it to the [normal import checking logic](https://webassembly.github.io/gc/core/valid/matching.html#match-globaltype).

No disagreement there. And the implementation complexity delta is minimal; within that minimal delta, an exact valtype match is ever so slightly simpler to implement and faster to execute.

I think I see the difference in perspective now: I'm coming from the function imports we're specifying in this proposal, where we clearly say "if you import a function `"wasm:js-string" "cast"`, and it doesn't have exactly the right signature, we'll throw a `LinkError`". By analogy, I was assuming that we're also restricting the type of magic string constant imports: "if you import something in module `"'"`, and it isn't a global with exactly the right type, we'll throw a `LinkError`". And for this "right type", we can basically pick any constraint we want, just like for the function signatures. The regular subtyping-based import checking logic does the rest, we certainly shouldn't change that. I consider it simplest to pin down one particular type, rather than allowing a degree of freedom that doesn't make a difference in practice. Again, there's an analogy to the functions: we could have specified `"fromCharCodeArray"` to return "either `(ref extern)` or `externref`", but there was no reason to allow both, and specifying/checking one exact required signature is simpler, so we did that. Why would we do something different for the imported globals?

> any module instantiated with magic imports could be instantiated with normal imports instead

Yes, that's the case either way. (The "and vice versa" part obviously only applies to very specific imports for the time being, given that we're specifying only very particular cases here. For example, there is currently no proposed way to have magic imports with any of the numeric types.)

I continue to be unconvinced by the observation that one could _hand-write_ a .s file that LLVM will _almost_ turn into a magic-string-imports compatible module. Realistically, to make use of imported string constants, a toolchain will need to be taught about this proposal (and likely all of WasmGC: we have previously discussed at length that it's difficult for non-GC languages to make use of ref-typed GC'ed strings), and then it's not hard to teach it about non-nullable types as well. In fact, it's likely even required anyway, seeing that many of the functions in this proposal involve non-nullable types in their signatures.

Reply:
I do think it makes more sense to have the rules for magic imported globals match the rules for non-magic imported globals rather than trying to make them match the rules for imported functions. FWIW, there's no difference between the rules for normal imported functions and imported string functions.

Reply:
I agree that the rules should follow those for regular imports. As for function imports, regular subtyping extends to these, see e.g. [this test](https://github.com/WebAssembly/gc/blob/9d2d6770914316193787ae64a61c697ef692a79a/test/core/gc/type-subtyping.wast#L497). But the function types specified in the String API have no declared supertypes, so this flexibility cannot be observed for them.

Reply:
> Why are globals' types more permissive than function signatures?

Because function types are not value types but "composite types", on which subtyping is only possible when it is pre-declared in the respective type definitions. For example, if the JS API declared "fromCharCode" to have type $t1 and that was defined as
```
(type $t0 (sub (func (param i32) (ref null extern))))
(type $t1 (sub $t0 (func (param i32) (ref extern))))
```
then you could indeed import it with either type $t0 or $t1. But that's not how this proposal defines its type, so there is not supertype you could use.

In contrast, subtyping on global types is determined directly by their value type's subtyping.

This are exactly the rules for regular imports.

(I don't know whether or not type errors for magic globals are supposed to be delayed under this proposal, or why that should be so.)

Reply:
To clarify, I'm not asking how subtyping works, and I never suggested that subtyping rules should be changed or special-cased for any kind of import.

My question is: this proposal specifies certain functions for which opt-in special handling can be requested, and they must have exactly one type for that to succeed. We could have allowed more types there (e.g. the subtyping you just mentioned), but there was no reason for that, so we kept it simple. So far so good.
Now this proposal also specifies certain globals for which opt-in special handling can be requested, and currently allows them to have one of several possible types. Why? To my eyes it's inconsistent to be less strict for them. Sure, we _could_ allow more types for them, but there is no reason to do so. The fact that it's easy to express the inconsistency on the Wasm level doesn't provide motivation _why we should have inconsistency_, especially when the fully consistent alternative is just as easy to express.

Reply:
My interpretation is that magic string globals conceptually have a unique type, namely `(ref extern)`. It just happens to be implied by the standard linking rules that you can then also import them with any supertype. The only such supertype currently is `(ref null extern)`. The same logic applies for functions, except that their types happen to have no other supertypes.

This seems maximally consistent with the existing rules.

Reply:
I think I see where @jakobkummerow is coming from here. There are two independent phases: first the names and types of imports are examined to determine what globals (constructed with string values) and functions are automatically added to the supplied imports, and second the normal instantiation process happens with all of its normal import checking rules.

My position has been that it would be useful to have the property that any instantiation that explicitly imports `WebAssembly.Global`s constructed with (valid unicode) strings should be able to use magic imports for those strings instead. That requires the first phase (which is not doing instantiation!) to produce a `Global` for any appropriately named global import for which instantiation with that string should succeed. Notably, this would require allowing magic imports to work with `anyref`, `(ref any)`, `i32`, and `f32` globals in addition to `externref` and `(ref extern)`, since globals of any of those types can be constructed from string values.

@jakobkummerow's position is that the rules in the first phase have nothing to do with the rules in the second phase, which remain unchanged, so we might as well keep things simple by only producing globals when the imports have precisely the `(ref extern)` type.

What if we simplified all this even further and had the first phase try to create `Global`s based only on the names of imported globals and not their types? The types would simply be passed along to the existing routine for constructing `WebAssembly.Global` with all its existing error checking. That achieves the property I've been advocating for with even fewer new checks than if we only allowed magic imports to work for `(ref extern)` globals.

Alternatively, we could have the first phase produce `Global`s again based only on the names, but rather than passing the imported global's value type to the `WebAssembly.Global` constructor, they would always pass `(ref extern)`, producing a later validation error if the actual type of the import was anything other than `externref` or `(ref null extern)`. This is closer to what @rossberg described.

(cc @SPY in case you're interested)

Reply:
@tlively , thanks for bridging the gap in perspectives here. I think you've captured my point.

Two comments:
I'm not sure I would call an extension of the principle to other types a "further simplification"; it seems like a pretty large extension of scope of the proposal. IIUC, you're saying that `(global (import "'" "42") i32)` should be automatically populated with `42` as imported value? One issue that comes to mind is that the Wasm types used at the boundary aren't very expressive for now. We're mapping `externref` to strings only by context (name of the `importedStringConstants` flag, name of this proposal, choice of `"'"` as module name). As you know, `externref` and `anyref` can be anything. For `(global (import "'" "42") anyref)`, how should the engine figure out whether you want `42` as a number (which is a perfectly valid `anyref` in a JS embedding) or `"42"` as a string? (Also, if the Wasm module knows that it wants an immutable `42`, why would it _import_ it with magic that only works in JS embeddings, as opposed to simply defining it as a non-imported global, or even inlining and constant-folding/-propagating it?)

The "first phase" in your description happens during (or right after) decoding. So from an implementation perspective, it's on a background thread and can't allocate on the heap. It might _conceptually_ create `Global`s, it doesn't actually create anything. Think of it as roughly:
```
for (Import i : module.imports) {
  if (import.module_name == "'") {
    if (MatchesJsStringBuiltinsRequirements(import)) {
      import.auto_populated = true;
    } else {
      return LinkError;
    }
  } else if (import.module_name == "wasm:js-string") {
    // Similar checks here.
  }
}
return OK;
```
Whereas the "second phase" is the actual instantiation:
```
for (Import i : module.imports) {
  if (import.auto_populated) {
    import.value = new String(...);
    // or, if it's one of the functions:
    import.value = new JSFunction(...);
  } else if (provided_imports.has(import.name, import.type)) {
    import.value = provided_imports.get(import);
  } else {
    return LinkError;
  }
}
```
Since the first phase's requirements are a fully compatible subset of regular Wasm rules, implementations don't need to perform any actual type validation for auto-populated imports in the second phase; the resulting behavior is the same as if such checks had been performed.

Reply:
I agree with @jakobkummerow that generalising this mechanism beyond strings raises various questions that have no satisfying answers, while also being of more dubious utility. So I'd vote for restricting it to JS strings. That also has the big advantage that it allows synthesising these imports purely based on name, without any dependency on types.

Reply:
Following up on myself: If we wanted to generalise this beyond strings, then that would require the ability to indicate the intended _JS_ type of the value, the Wasm target type is not good enough.

One natural way to do that would be to actually parse the import name as a JS literal, assigning it the same semantics as it would when evaluated in JS, which then is used to try satisfying the import in the standard manner. But to make that work, strings would require extra quotes, as in `"\"42\"", while `"42"` would just be the number. You'd also have to escape some characters to satisfy JS string syntax. It would further raise the question what extent of JS syntax to allow, e.g., is `"[42, 43]"` okay as well? Where do we stop? Overall, I doubt that's all worth it.

Another option would be to distinguish by different pseudo module names, e.g., `"'"` for strings, but `"."` for numbers. But again, I don't think that extension is worth it, especially since strings are the only known use case for this.

Reply:
@jakobkummerow 

> Now this PR in its current form adds the following behaviors IIUC:
> 
>     (global (import "'" "foo") i32) → OK for WebAssembly.compile, CompileError later during WebAssembly.instantiate. What's the reason for the inconsistency? Why not check during WebAssembly.compile? Why CompileError and not LinkError?
>     (global (import "'" "foo") (ref extern)) → OK (uncontentious)
>     (global (import "'" "foo") (ref null extern)) → OK. Why are globals' types more permissive than function signatures?
>     (global (import "'" "foo") (mut (ref extern)) → OK for WebAssembly.compile, CompileError later during WebAssembly.instantiate. So we do have some extra restrictions for these globals, the principle isn't simply to leave them unconstrained. Why is immutability enforced but non-nullability is not?
> 

For the first and last cases here, I had intended for these to fail WebAssembly.compile with an eager CompilerError, just like when there is a mismatch with functions. That was the purpose of "This check is eager, resulting in a compile error if it fails." line in this PR on line 176.

@tlively 

>  I think I see where @jakobkummerow is coming from here. There are two independent phases: first the names and types of imports are examined to determine what globals (constructed with string values) and functions are automatically added to the supplied imports, and second the normal instantiation process happens with all of its normal import checking rules.

I think the two differences I have is that I was intending for the import names alone to be what determines for global/function type is provided to the 'supplied imports'. Each builtin function/global has a type determined by the spec, the import name does a 'builtin lookup' with it. And then I was intending for the import matching that normal instantiation would do for these 'supplied imports' to happen during compile time instead.

We could have a model where the import names _and_ types are part of the 'builtin lookup', but I'd like to hear the benefit of this. In this case for 'imported strings', it might result in a slightly simpler check when compiling, but I don't think that's worth the irregularity.

If we were to introduce a type import for JS strings (as a subtype of externref), this model would extend naturally so that we could redefine the imported string globals as `(global (ref $JSString))` and the normal subtyping rules would ensure that old modules that imported `(ref extern)` would continue to work.

Reply:
AFAICS, there is no inconsistency in what @eqrion describes. The difference in subtyping between global and function types is completely consistent with regular imports and just a consequence of the existing requirement to explicitly pre-declare subtyping between the latter, see my earlier explanation.

Reply:
@jakobkummerow I agree the overview could go into more detail here. I'll push a commit here to specify things better.